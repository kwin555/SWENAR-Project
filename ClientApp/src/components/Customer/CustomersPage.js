import React, { Component } from "react";
import { AgGridReact } from "ag-grid-react";
import CustomerForm from "./CustomerForm";
import { URL } from "./CustomerConstants";
import "ag-grid-community/dist/styles/ag-grid.css";
import "ag-grid-community/dist/styles/ag-theme-balham.css";

//default styling the customers react data grid
const style = {
  height: "500px",
  width: "610px",
  justifyContent: "center"
};

export class CustomersPage extends Component {
  static displayName = CustomersPage.name;
  //default state of the compoment is empty customers
  constructor(props) {
    super(props);
    this.state = {
      customers: []
    };
  }
  // used to fetch all of the customers from the end point
  fetchCustomers = () => {
    fetch(URL)
      .then(res => res.json())
      .then(data => {
        console.log(data);
        this.setState({ customers: data });
      })
      .catch(error => {
        console.log(error);
      });
  };
  // when a page loads fetch the customer information
  componentDidMount = () => {
    this.fetchCustomers();
  };
  // if changes are made to customers update the datagrid.
  componentDidUpdate = ({ customers: prevCustomers }) => {
    const { customers } = this.state;
  };

  // update helper function to update name change.
  handleNumName = (Num, Name) => {
    let customer = {
      id: this.state.customers[this.state.customers.length - 1].id + 1,
      name: Name,
      number: Num
    };
    this.setState({
      customers: [...this.state.customers, customer]
    });
  };

  // properties of each customer
  columns = [
    {
      field: "id",
      name: "ID",
      cellRenderer: params => {
        var link = document.createElement("a");
        link.href = "/customerpage/" + params.value;
        link.innerText = params.value;
        return link;
      }
    },
    { field: "name", name: "Name" },
    { field: "number", name: "Number" }
  ];

  render = () => {
    // Rendering out hhe customers React data grid.
    return (
      <div>
        <div>
          <h3>Customers</h3>
          <CustomerForm handleNumName={this.handleNumName} />
          <hr />
        </div>
        <div className='ag-theme-balham' style={style}>
          <AgGridReact
            columnDefs={this.columns}
            rowData={this.state.customers}
          ></AgGridReact>
        </div>
      </div>
    );
  };
}
