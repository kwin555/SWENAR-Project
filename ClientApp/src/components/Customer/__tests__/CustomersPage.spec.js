import React from "react";
import CustomersPage from "../CustomersPage";
import { shallow, mount } from "enzyme";
import axios from "axios";
import fetchMock from "fetch-mock";

describe("CustomerPage tests", () => {
  it("shows customer page with things rendered", () => {
    const props = {
      customers: [{ Name: "Kevin", number: "23423" }],
    };

    const customersPage = shallow(<CustomersPage {...props} />);
    expect(customersPage).toMatchSnapshot();
  });

  it("should handleNumName", () => {
    const props = {
      customers: [{ id: 1, Name: "Kevin", number: "23423" }],
    };

    const customersPage = shallow(<CustomersPage {...props} />);

    customersPage.setState({
      customers: [{ id: 1, Name: "Kevin", number: "23423" }],
    });
    customersPage.instance().handleNumName("1234", "Zach");

    expect(customersPage.state().customers).toEqual([
      { id: 1, Name: "Kevin", number: "23423" },
      { id: 2, name: "Zach", number: "1234" },
    ]);
  });

  it("should return columns", () => {
    const props = {
      customers: [{ id: 1, Name: "Kevin", number: "23423" }],
      match: {
        params: 3,
      },
    };

    const customersPage = shallow(<CustomersPage {...props} />);

    const columns = customersPage.instance().columns();

    expect(columns[0].cellRenderer({ value: 3 }).toString()).toEqual(
        'http://localhost/customerpage/3'
    );
  });
});
