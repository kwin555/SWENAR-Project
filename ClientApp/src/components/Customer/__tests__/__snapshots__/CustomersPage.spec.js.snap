// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerPage tests shows customer page with things rendered 1`] = `
<div>
  <div>
    <h3>
      Customers
    </h3>
    <CustomerForm
      handleNumName={[Function]}
    />
    <hr />
  </div>
  <div
    className="ag-theme-balham"
    style={
      Object {
        "height": "500px",
        "justifyContent": "center",
        "width": "610px",
      }
    }
  >
    <AgGridReact
      columnDefs={[Function]}
      rowData={Array []}
    />
  </div>
</div>
`;
